### ISSUE TITLE
Corrupted output from inception_eval

### ISSUE THREAD

**Issue by enizhibitsky** [tensorflow/models#50]
I'm trying to get out predictions with my fine-tuned model using a script similar to `inception_eval.py`.

I don't have true labels for it so I want to save both initial filenames and logit predictions.

I modified `image_processing` so it returns filenames too and I can use it like that:
```python
images, labels, filenames = image_processing.inputs(dataset)
```
The problem is when I add this to the while-loop:
```python
while step < num_iter and not coord.should_stop():
  paths, targets, preds = sess.run([filenames, labels, logits])
  for path in paths:
    print(path)
```
I got about 0.1% duplicated values (and 0.1% missing thereby).

Seems like thread-safety issue somewhere in the modules or I'm missing something.

**Comment by girving** [comment]
@enizhibitsky: I'd check if your `image_processing.input` is still using a `RandomShuffleQueue`. :)

**Comment by aselle** [comment]
Automatically closing because there was no response. Please reopen if it is still an issue.

**Comment by enizhibitsky** [comment]
@girving the time i opened this issue i didn't have any `Queue`'s yet :)

Now I have up-to-date `master`'s of both `tensorflow` and `models` repositories but it is still the issue.

I modified `image_processing` to return the original filenames:
```python
'image/filename': tf.FixedLenFeature([], dtype=tf.string,
                                     default_value=''),
<...>
images_and_labels_and_filenames = []
<...>
image_buffer, label_index, bbox, _, filename = parse_example_proto(
<...and so on...>
```

And made a minimal case for this:
``` python
"""A binary to predict Inception on the custom data set.
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from datetime import datetime
import math

import numpy as np
import tensorflow as tf

from inception import image_processing
from inception.custom_data import CustomData


FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_integer('num_examples', 32000,
                            """Number of examples to run. Note that the test """
                            """Custom dataset contains 32000 examples.""")

if __name__ == "__main__":
  dataset = CustomData(subset='validation')
  num_classes = dataset.num_classes()

  paths = np.zeros(FLAGS.num_examples, dtype=int)

  with tf.Graph().as_default():
    _, _, filenames = image_processing.inputs(dataset)

    with tf.Session() as sess:
      coord = tf.train.Coordinator()
      try:
        threads = tf.train.start_queue_runners(sess, coord, daemon=False)

        num_iter = int(math.ceil(FLAGS.num_examples / FLAGS.batch_size))
        step = 0

        while step < num_iter and not coord.should_stop():
          path = sess.run([filenames])[0]

          from_idx = FLAGS.batch_size * step
          last_idx = FLAGS.batch_size * step + np.min([FLAGS.batch_size, len(paths[from_idx:])])

          # path/to/image/123.jpg -> 123
          to_id = lambda s: int(s.split('/')[-1][:-4])
          paths[from_idx:last_idx] = map(to_id, path)[:last_idx - from_idx]

          step += 1
          if step % 100 == 0:
            print('%s: [%d/%d] (%d)' % (datetime.now(),
                                        last_idx,
                                        FLAGS.num_examples,
                                        len(np.unique(paths[:last_idx]))))

      except Exception as e:
        coord.request_stop(e)

      else:
        coord.request_stop()
        coord.join(threads, stop_grace_period_secs=10)
        print('%s: Unique: [%d/%d]' % (datetime.now(),
                                       len(np.unique(paths[:last_idx])),
                                       FLAGS.num_examples))

```
And the output looks like that:
```
2016-07-25 13:31:16.611949: [3200/32000] (3200)
2016-07-25 13:31:24.239994: [6400/32000] (6400)
2016-07-25 13:31:31.909083: [9600/32000] (9600)
2016-07-25 13:31:39.551816: [12800/32000] (12800)
2016-07-25 13:31:47.222275: [16000/32000] (16000)
2016-07-25 13:31:54.957032: [19200/32000] (19200)
2016-07-25 13:32:02.639116: [22400/32000] (22400)
2016-07-25 13:32:10.489635: [25600/32000] (25600)
2016-07-25 13:32:18.050272: [28800/32000] (28800)
2016-07-25 13:32:25.817593: [32000/32000] (31999)
<...>
2016-07-25 13:32:25.835171: Unique: [31999/32000]
```
Sometimes it returns 32000 uniques, sometimes even less then 31999.

**Comment by girving** [comment]
I'm having trouble reading the meaning of the `<...>` sections.

**Comment by enizhibitsky** [comment]
Updated my comment, added a part of `git diff` to be more clear. `<...>`'s in the output are just non-informative debug messages. The point is in uniques count that is less then expected.

**Comment by girving** [comment]
If you look at `image_processing.inputs`, it should be iterating through the files in `dataset.data_files()` one at a time.  Is it doing it in the correct order?

**Comment by enizhibitsky** [comment]
I can't look after iterating through `dataset.data_files()` as for now (lots of references through ops python files confuse me), but I've made some further investigations — the issue can be reproduced even on a "dataset" with 8 files in 4 directories with an output like (filenames printing added):
```
% bash case.sh 2>/dev/null
['3.jpg']
['1.jpg']
['6.jpg']
['7.jpg']
['5.jpg']
['8.jpg']
['3.jpg']
['8.jpg']
2016-07-27 00:08:30.014559: Unique: [6/8]

% bash case.sh 2>/dev/null
['3.jpg' '7.jpg' '8.jpg' '1.jpg']
['6.jpg' '5.jpg' '4.jpg' '3.jpg']
2016-07-27 00:10:53.294479: Unique: [7/8]

% bash case.sh 2>/dev/null
['3.jpg' '7.jpg' '8.jpg' '6.jpg' '1.jpg' '4.jpg' '5.jpg' '3.jpg']
2016-07-27 00:12:01.431401: Unique: [7/8]
```
Most of the times 3.jpg or 8.jpg are doubled while some of other (different) images are absent.

By the way, the issue dissapears with `num_preprocess_threads=1` and its div by 4 checking disabled, and the output becomes constant, but I remember this didn't help me with 32000 images.

**Comment by girving** [comment]
So this behavior is expected with more than one thread.  You are cycling through the input, and one thread is managing to reach the end of the list of filenames, wrap around, and stick in the first filename before another thread manages to produce its result.  It's not "thread safety", just normal asynchronous behavior.

Are you sure this isn't also the cause of the full issue?

If you want it to not do this, you'll need to pass `num_epochs=1` to `string_input_producer` inside `image_processing.batch_inputs`.  This will prevent it from wrapping around.

**Comment by girving** [comment]
Closing for now; maybe my suggestions were correct?  I'm happy to reopen if more details emerge.
